qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,261 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""THC SELECT and PREPARE"""
+from functools import cached_property
+from typing import Dict
+
+import numpy as np
+from attrs import frozen
+
+from qualtran import Bloq, BloqBuilder, Register, Signature, SoquetT
+from qualtran.bloqs.arithmetic import (
+    EqualsAConstant,
+    GreaterThanConstant,
+    LessThanConstant,
+    LessThanEqual,
+)
+from qualtran.bloqs.basic_gates import Hadamard, Ry, Toffoli, XGate, ZGate
+from qualtran.bloqs.chemistry.contiguous_register import ContiguousRegister
+from qualtran.bloqs.controlled_bloq import ControlledBloq
+from qualtran.bloqs.multi_control_multi_target_pauli import MultiControlPauli
+from qualtran.bloqs.on_each import OnEach
+from qualtran.bloqs.qrom import QROM
+from qualtran.bloqs.swap_network import CSwapApprox
+
+
+@frozen
+class UniformSuperpositionTHC(Bloq):
+    r"""Prepare uniform superposition state for THC.
+
+    $$
+        |0\rangle^{\otimes 2\log(M+1)} \rightarrow \sum_{\mu\le\nu}^{M} |\mu\rangle|\nu\rangle + \sum_{\mu}^{N/2}|\mu\rangle|\nu=M+1\rangle,
+    $$
+
+    where $M$ is the THC auxiliary dimension, and $N$ is the number of spin orbitals.
+
+    The toffoli complexity of this gate should be 10 * log(M+1) + 2 b_r - 9.
+    Currently it is a good deal larger due to:
+        1. inverting inequality tests should not need more toffolis.
+        2. We are not using phase-gradient gate toffoli cost for Ry rotations
```

Comment:
I'd suggest opening (and linking) an issue on github so we can keep track of this

Reply:
Done.